### PR TITLE
Insert regmove instructions in new blocks dedicated to hold the diversions.

### PR DESCRIPTION
**Pull request by nbp**
This work merges #844 (split critical edges) and #883 (add entry register diversions), and modifies the coloring and location verifier to prevent register move instructions to be added between a conditional branch and the jump instructions which follows it.

Keeping this pull request as a draft until both pull request from which it depends on are not merged.

### CODE REVIEW

File: filetests/regalloc/x86-regres.clif [hunk 1]
```diff
@@ -24,17 +25,19 @@ ebb2(v4: i32, v5: i32, v7: i32):
     v8 = iconst.i32 -1
     ; v7 is killed here and v9 gets the same register.
     v9 = iadd v7, v8
-    ; check: v9 = iadd v7, v8
+    ; check:   v9 = iadd v7, v8
```
Comment (bjorn3) [inline filetests/regalloc/x86-regres.clif]: Why these extra space? `check` should trim the input.

Comment (nbp) [reply filetests/regalloc/x86-regres.clif]: This is to have a 2 spaces indentation of instructions compared to basic blocks labels. These should not matter when testing.

**Comment by nbp** [comment]
Here are the embenchen results, where the first column is cranelift with basic-blocks feature enabled, and the second  column is cranelift without basic-blocks feature enabled (excluding branch splitting $844 (without #887 optimization), and not forwarding diversions set to the next block (based on #883 )).

```
# mode=cranelift+cranelift, runs=31, problem size=default
# Lower score is better
box2d           1205    1306    0.923
bullet          1750    1773    0.987
conditionals    422     456     0.925
copy            773     786     0.983
corrections     1078    1060    1.017
fannkuch        3023    3255    0.929
fasta           943     884     1.067
fib             778     787     0.989
ifs             231     246     0.939
binarytrees     6045    6084    0.994
memops          1785    2163    0.825
primes          1159    1150    1.008
raybench        2458    2540    0.968
rust-fannkuch   3071    3097    0.992
skinning        833     880     0.947
zlib            1679    1740    0.965
```

File: cranelift-codegen/src/regalloc/coloring.rs [hunk 2]
```diff
@@ -183,6 +187,54 @@ impl<'a> Context<'a> {
...
+                            "Divert set is non-empty after the terminator."
+                        );
+                    }
+                    debug_assert_eq!(
+                        self.cur.next_inst(),
+                        None,
+                        "Unexpected instruction after a branch group."
+                    );
+                } else {
+                    debug_assert!(opcode.is_terminator());
```
Comment (lars-t-hansen) [inline cranelift-codegen/src/regalloc/coloring.rs]: Better to panic here?

File: cranelift-codegen/src/regalloc/coloring.rs [hunk 3]
```diff
@@ -183,6 +187,54 @@ impl<'a> Context<'a> {
...
+                    let target = match self.cur.func.dfg.analyze_branch(branch) {
+                        NotABranch | Table(_, _) => panic!(
+                            "unexpected instruction {} after a conditional branch",
+                            self.cur.display_inst(branch)
+                        ),
+                        SingleDest(ebb, _) => ebb,
+                    };
+
+                    // We have a single branch with a single target, and an EBB with a single
+                    // predecessor. Thus we can forward the divertion set to the next EBB.
```
Comment (lars-t-hansen) [inline cranelift-codegen/src/regalloc/coloring.rs]: Spelling: "diversion".

File: cranelift-codegen/src/regalloc/coloring.rs [hunk 4]
```diff
@@ -183,6 +187,54 @@ impl<'a> Context<'a> {
...
+                            target,
+                            self.divert.display(&self.reginfo)
+                        );
+                    } else {
+                        debug_assert!(
+                            self.divert.is_empty(),
+                            "Divert set is non-empty after the terminator."
+                        );
+                    }
+                    debug_assert_eq!(
```
Comment (lars-t-hansen) [inline cranelift-codegen/src/regalloc/coloring.rs]: Better to panic here?

**Pull request merged**